Kdenlive/mlt: enable glaxnimate to add vectorial animations

tobiasBora opened the pull request:
###### Description of changes

Fix https://github.com/NixOS/nixpkgs/issues/209923

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

@turion @goibhniu

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/kdenlive-missing-mlt-module-glaxnimate/24486/6

turion commented on pkgs/development/libraries/mlt/qt-5.nix:
```diff
@@ -33,7 +35,8 @@ mkDerivation rec {
     rev = "v${version}";
-    sha256 = "sha256-Y7lbfwA0lkQB3PjYQIQaQ0BeXGcgyCmMnDqqZJ8zUaA=";
+    sha256 = "sha256-rio8HpSa3CpyyrrbaaEmHzM+xJgCl1CCAlVs/9IXq9U=";
+    fetchSubmodules = true; # For now only contains glaxnimate… not sure how/if it's possible to use directly the packaged glaxnimate.
```
What happens if you don't do this?

tobiasBora replied:
It fails to find some `*.in.hpp` files downloaded by the submodule.

SuperSandro2000 replied:
Can we symlink our glaxnimate to that location instead?

tobiasBora replied:
You mean the final derivation? If yes, no because glaxnimate's output derivation is just a binary + some share files (icons, plugins, python scripts…), while here it tries to use internal functionality of glaxnimate that are not exposed as libraries.

chayleaf replied:
can you use `glaxnimate.src` then?

tobiasBora replied:
I have to admit that I don't see much advantage of using `glaxnimate.src` as an upgrade to glaxnimate could easily break the current package, for instance if glaxnimate removes or rename the C functions used in this plugin… (which will certainly arrive as they plan to refactor the code of the library) And I don't see much advantage over fetching the module, anI missing something?

turion replied:
I think fetching it as a submodule is fine as long as it isn't packaged better.

turion commented on pkgs/development/libraries/mlt/qt-5.nix:
```diff
@@ -55,7 +58,11 @@ mkDerivation rec {
-  ];
+  ] ++ lib.optionals enableGlaxnimate ([
+    # Enable glaxnimate https://github.com/NixOS/nixpkgs/issues/209923;
+    glaxnimate
+    # Not sure why, but mlt needs to compile itself glaxinate, instead of picking the parent project (that, admitingly, only contains the binary)
```
These two lines are somewhat contradictory: Either you rebuild glaxnimate here, or you use the build input.

turion commented on pkgs/development/libraries/mlt/qt-5.nix:
```diff
@@ -55,7 +58,11 @@ mkDerivation rec {
+  ] ++ lib.optionals enableGlaxnimate ([
+    # Enable glaxnimate https://github.com/NixOS/nixpkgs/issues/209923;
+    glaxnimate
+    # Not sure why, but mlt needs to compile itself glaxinate, instead of picking the parent project (that, admitingly, only contains the binary)
```
Maybe it's possible to build glaxnimate in a different way such that it produces a library? Maybe the glaxnimate derivation can be changed to contain several outputs?

tobiasBora replied:
Could be, but I found no easy way reading at the documentation/cmake file, and I'm not extremely fluent with cmake files nor the internal details of these prolects.

turion replied:
I have next to no idea about cmake, or C/C++ building in general :/ I think we should rather stall this PR here until there is some feedback from the mlt folks, or another idea by someone else from nixpkgs.

SuperSandro2000 replied:
if glaxnimate always requires it's buildInputs to be used as a library we should propagate them instead.

tobiasBora replied:
Well I use `glaxnimate.buildInputs` only because `mlt` tries to compile glaxnimate from scratch, and I guess it's cleaner to bring all dependencies this way rather than copy pasting the 10 dependencies of glaxnimate.

tobiasBora replied:
@turion so I got some feedbacks from MLT: they won't do anything themself but are open to PR. Regarding glaxnimate, they want to refactor the project to provide explicit small reusable libraries, but so far they had no time to do it. So for now I don't think they want to change their build system to include libraries that won't be available anymore after the refactorization.

SuperSandro2000 commented on pkgs/development/libraries/mlt/qt-5.nix:
```diff
@@ -65,7 +72,7 @@ mkDerivation rec {
     "-DCMAKE_SKIP_BUILD_RPATH=ON"
     "-DMOD_OPENCV=ON"
-  ];
+  ] ++ lib.optional enableGlaxnimate "-DMOD_GLAXNIMATE=ON"; # Enable glaxnimate https://github.com/NixOS/nixpkgs/issues/209923;
```
```suggestion
  ] ++ lib.optional enableGlaxnimate "-DMOD_GLAXNIMATE=ON";
```
that is clear from the line already and I am not sure why we link the issue here which can be closed after this is merged.

SuperSandro2000 reviewed (commented):
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

tobiasBora commented:
So I got an answer by glaxnimate's team: https://gitlab.com/mattbas/glaxnimate/-/issues/545. They say they plan to modify their software to cut it in smaller, reusable libraries, but they had no time to do it so far. So I guess for now it's better to use submodules rather than spending time on writing patches that would eventually be unnecessary.

If you are good with that, I can remove `glaxnimate` from the list of dependencies, squash & modify the commit name once @SuperSandro2000 gives me the appropriate name to give to the commit, and we can merge if everything is good for you.

SuperSandro2000 commented:
> Do I need to write one commit per program?

Yes

> So I guess for now it's better to use submodules rather than spending time on writing patches that would eventually be unnecessary.

If the required work is just installing the so, even if it is done via a cp command and linking against that, then that is preferred. Package collections, like nixpkgs, want to avoid vendoring as much as possible because it makes having a single source for a package with all patches impossible.

HiWhatName commented:
Any updates on this?

tobiasBora commented:
> If the required work is just installing the so, even if it is done via a cp command and linking against that, then that is preferred.

@SuperSandro2000 I understand that Nix tries to limit third party sources, but the software needs the source since glaxnimate provides no library, so I don't think a .so file is even produced (but I did it a while ago so I might remember poorly). I have no idea how to change the whole compilation process in order to avoid this dependency on the source (and I don't have time to spend one week on this), so if you know how to do it feel free to… until then people won't be able to use glaxnimate in KDEnlive.

tobiasBora commented:
> Where are we stuck currently?

The code works (or at least used to work, I've not tested recently), I am just waiting for @SuperSandro2000 to accept the fact that I rely on a git submodule following what is done upstream. Since glaxnimate provides no API/.so library for now (it might [come later if they find people with good enough cmake expertise](https://gitlab.com/mattbas/glaxnimate/-/issues/545)), I see no other solution.

turion commented:
@tobiasBora see https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions for naming conventions for the commits. I think ideally you would split the commit into two, with these names:

```
kdenlive: enable glaxnimate
mlt: enable glaxnimate
```

turion commented:
@ofborg build kdenlive

turion commented:
There is something wrong with the build: 

```
building '/nix/store/qw3y9318g1v0i0s8x90dav38ijg4qva6-mlt-7.22.0.drv'...
Running phase: qtPreHook
Error: wrapQtAppsHook is not used, and dontWrapQtApps is not set.
error: builder for '/nix/store/qw3y9318g1v0i0s8x90dav38ijg4qva6-mlt-7.22.0.drv' failed with exit code 1;
       last 2 log lines:
       > Running phase: qtPreHook
       > Error: wrapQtAppsHook is not used, and dontWrapQtApps is not set.
```

https://github.com/NixOS/nixpkgs/pull/209941/checks?check_run_id=20524628144

I can reproduce the build error locally.

turion commented:
It must somehow be related to this part:

```
  ] ++ lib.optionals (qt != null) [
    qt.wrapQtAppsHook
];
```

But it builds on master. I don't understand what change creates the error.

turion commented on pkgs/development/libraries/mlt/default.nix:
```diff
@@ -94,7 +100,7 @@ stdenv.mkDerivation rec {
   ] ++ lib.optionals enableSDL2 [
     SDL2
-  ];
+  ] ++ lib.optionals enableGlaxnimate glaxnimate.buildInputs;
```
I think this doesn't quite work as intended anymore. Maybe the issue is that this unconditionally adds `qtbase` and other qt packages, but `mlt` can only deal with them when `qt != null` in the inputs. Maybe throw an error if `enableGlaxnimate = true && qt == null`?

tobiasBora commented:
Oh interesting… I was also getting this error when compiling mlt alone, but I thought that it was expected since `qt` was not set, thinking that mlt was only compilable with a non-empty qt. Since kdenlive was able to compile `mlt` I thought that mlt was getting sent qt somehow. But now I'm very confused: why is `mlt` failing to build while `ffmpeg = mlt.override {ffmpeg-full}` is not?

tobiasBora commented:
Oh, ok, now I get why mlt in kdenlive builds fine: in `pkgs/top-level/qt5-packages.nix`, there is 
```
  mlt = pkgs.mlt.override {
    qt = qt5;
  };
```
so KDEnlive is using the one with Qt, explaining why it is not failing to build.

When I have a bit more time, I will try to see if qt is really needed, and if it is needed I will maybe enable glaxnimate only if qt is not null (throwing an error would break the current version, so I'm thinking it might not be the best idea).

ShalokShalom commented:
Just as a heads up, this is still the case in the unstable build with Qt6 (kdePackages.kdenlive)

tobiasBora commented:
Ok thanks. I need to take a bit of time to fix this issue. The next month will be rather busy for me, I'll see if I can find some time before but it's definitely in my todo list.

tobiasBora committed:
mlt: add glaxnimate support

tobiasBora commented:
This version should work now (let's wait maybe for the tests to finish first). By just specifying the only actually needed library, we get rid of the mlt error (actually, mlt seems to automatically disable glaxnimate support if qt is disabled, without providing any error).

tobiasBora commented:
Seems like all tests passes, are we good now?

SuperSandro2000 commented on pkgs/applications/kde/kdenlive/default.nix:
```diff
@@ -80,13 +82,14 @@ mkDerivation {
+  patches = [ ./dependency-paths.patch ] ++ lib.optional enableGlaxnimate ./dependency-paths-glaxnimate.patch;
 
-  inherit mediainfo;
+  # Needed to replace @glaxnimate@ by its path
+  inherit mediainfo glaxnimate;
```
Doesn't this need to be done with substituteAll on the patch file?

tobiasBora replied:
Sure, but `substituteAllInPlace` is already part of Kdenlive's derivation as this file is already patched similarly for other binaries.

SuperSandro2000 replied:
Can we make this explicit? Using substituteAllInPlace with ambient variables is kinda surpising.

tobiasBora replied:
What do you mean? substituteAllInPlace just expects these variables to be in the environment (whether this is good or not is discussed [here](https://github.com/NixOS/nixpkgs/issues/237216), but so far no agreement has been reached), and my goal was just to implement this fix following what kdenlive's package is already doing, as I don't want to risk adding bugs doing some sort of refactoring. Or do you mean to put them in `env = {}` to make it somewhat clearer?

chayleaf replied:
I agree, it would be cleaner to use the substituteAll builder on the patch itself. That said, I don't mind either way.

tobiasBora replied:
I added an `env` in case you find it clearer, hopefully it will not introduce side effects (but after a quick test it seems to work). Any larger refactoring should be done in a different PR.

chayleaf commented on pkgs/development/libraries/mlt/default.nix:
```diff
@@ -36,6 +36,9 @@
 , enableSDL2 ? true
 , SDL2
 , gitUpdater
+, glaxnimate
```
```suggestion
```

turion commented:
@tobiasBora Hi :) do you want to rebase? I'd like to merge it.

K900 commented:
This will only affect the soon-to-be-removed Qt5 version. The Qt6 version is in pkgs/kde/gear/kdenlive.

tobiasBora commented:
Argggggggggg We should really follow a policy to accept PR faster, the current system wastes the time and energy of reviewers & maintainers…

turion commented:
> Argggggggggg We should really follow a policy to accept PR faster, the current system wastes the time and energy of reviewers & maintainers…

Yes, absolutely agree. I only learned about the Qt6 version being a different file today as well.

Now I guess the glaxnimate fix have to be transported to the Qt6 version?

tobiasBora committed:
kdenlive: add glaxnimate support for Qt6

tobiasBora committed:
mlt: fix it to work with Qt6

tobiasBora commented:
Ok, the latest version works with Qt6. Compile with:
```
nix-build -A kdePackages.kdenlive
```

K900 commented on pkgs/kde/gear/kdenlive/default.nix:
```diff
@@ -12,6 +13,8 @@
   frei0r,
   fetchpatch,
+  glaxnimate,
+  enableGlaxnimate ? true,
```
Again, does this really need to be conditional?

tobiasBora replied:
Sorry, I missed your first message, you are right I can make itnon optional. I also realized (thanks oborg) that my new commit fails to compile flowblade, so wait before merging.

turion replied:
> that my new commit fails to compile flowblade

Is this issue resolved?

tobiasBora replied:
Fixed

tobiasBora committed:
mlt: fix bug when qt is not enabled

tobiasBora committed:
kdenlive: make glaxnimate non-optional

tobiasBora commented:
Really? You made sure to use the very latest version d27c900cb723f4c6f03941c7c0b33a92e6411032? (Qt6 used to have this bug on a previous commit because I was not compiling mlt for Qt6, but I just tested both Qt5 & 6 with this commit  and both work for me…). If yes, are both versions (Qt5 & 6) failing? If yes, you can maybe try to reset the configuration, maybe it saved some outdated configuration. For this, try to run:
```
rm .config/kdenlive*
```
(or maybe do a backup if you value this config)

turion commented:
I used the version you mentioned, and tested it by creating a nixos test VM with this config:

```nix
{
    imports = [
      ./common/x11.nix
    ];

    services.xserver.enable = true;
    environment.systemPackages = with pkgs; [
      kdePackages.kdenlive
    ];
  }
```
So the home directory is always pristine.

Running the configuration wizard shows no error message, so something is working :sweat_smile: still, "Create animation" is greyed out:

![image](https://github.com/NixOS/nixpkgs/assets/303489/e5ed6064-bccc-4d67-a00d-7e55dee5af92)

tobiasBora commented:
I tried to investigate a bit, and I think it is a bug in Kdenlive that occurs only the first time you start it (or if you start it after an order version). I get exactly the first issue (with also "add Title clip" greyed like you, which is also something that should not be greyed) if I do:
```
$ rm ~/.config/kdenlive*
$ ./result/bin/kdenlive
```
but if I just close and restart kdenlive, it works again!

turion commented:
> but if I just close and restart kdenlive, it works again!

True, can reproduce that! Apart from this weird behaviour (which I can observe both in XFCE and Gnome), it seems to work fine :) well done, fantastic! Thank you in particular for your stamina on this PR!

tobiasBora commented:
Great thanks! Are we waiting for something else to merge? If we don't do it now, I'm afraid to have to solve yet another merge conflict in X months ^^'

turion commented:
There was the one discussion about the optional argument `enableGlaxnimate`, but I think it is not critical.

tobiasBora commented:
This one is fixed, glaxnimate is now enabled unconditionally in Kdenlive (which makes sense since it is anyway required by kdenlive if we want no error at startup). Thanks!

K900 commented:
Respectfully, what the hell? How did this get merged in the current state? The implementation isn't consistent across Qt versions, none of the maintainers signed off on this, and now we'll have to do a follow-up PR because I don't want to have this in tree in this state.

K900 commented on pkgs/applications/kde/kdenlive/default.nix:
```diff
@@ -80,15 +81,19 @@ mkDerivation {
   # doesn't find them. See:
   # https://github.com/NixOS/nixpkgs/issues/83885
-  patches = [ ./dependency-paths.patch ];
+  patches = [ ./dependency-paths.patch ./dependency-paths-glaxnimate.patch ];
```
Why are these two patches separate?

tobiasBora replied:
Technically, it was mostly to save me 2h of debugging & testing & compiling while I was quite busy for other stuff. I tried before to merge the two patches manually, but it turned out that for a reason I can't explain my patch was not applied (but I did not realized it directly), making me lose quite some time, first to realize that my patch was not applied, and then to find a way to combine these patches appropriately + burn some CPU power to compile both versions and redo a session of tests etc… And I thought that this small additional patch was not such a bad thing compared with the time it would save me ^^

K900 replied:
So you've chosen to save your time now instead of saving everyone's time later?

tobiasBora replied:
Well, I'm not so sure if this will make anyone lose 2h later since it does not harm much to have two files instead of 1. But don't get me wrong, I do get your point, especially for important architecture decisions, where I'm the first one to care about having a clean base on which to build on… this was just my tradeoff between answering quickly to turion's request without getting late on the other things I need to do. Obviously this had not the effect I was hopping for, so I know what I'm gonna do tonight :-P

K900 commented on pkgs/applications/kde/kdenlive/default.nix:
```diff
@@ -80,15 +81,19 @@ mkDerivation {
     wrapGAppsHook3
+    glaxnimate
   ];
   # Both MLT and FFMpeg paths must be set or Kdenlive will complain that it
```
Comment is outdated now.

tobiasBora replied:
what do you mean? I never wrote this comment I think ^^

K900 replied:
Yes, but you changed what the patch does, so the comment needs updating as well.

K900 commented on pkgs/applications/kde/kdenlive/default.nix:
```diff
@@ -80,15 +81,19 @@ mkDerivation {
-  inherit mediainfo;
-  ffmpeg = ffmpeg-full;
-  mlt = mlt-full;
+  # Needed to replace @glaxnimate@ by its path
```
This comment does not explain what all the other things are for.

tobiasBora replied:
removed in https://github.com/NixOS/nixpkgs/pull/324552

K900 commented on pkgs/development/libraries/mlt/default.nix:
```diff
@@ -36,6 +36,8 @@
 , enableSDL2 ? true
 , SDL2
 , gitUpdater
+, enableGlaxnimate ? true
```
Does this need to be optional here?

tobiasBora replied:
Since the library creates a compilation flag to enable this or not, I think it makes sense to make it optional here… But to be honest I don't care so much.

K900 replied:
Not every optional dependency should be a compilation flag, or we'll create an unmaintainable combinatoric explosion of possible flags.

tobiasBora replied:
Alright, I don't mind to remove it then. I saw that you plan to work on it now, should I let you do it?

K900 replied:
I'll remove it in the 24.05.2 PR.

tobiasBora replied:
Ok thanks a lot!

K900 commented on pkgs/kde/gear/kdenlive/default.nix:
```diff
@@ -20,7 +23,7 @@ mkKdeDerivation {
       substituteAll {
         src = ./dependency-paths.patch;
-        inherit mediainfo mlt;
+        inherit mediainfo mlt glaxnimate;
```
This does nothing

tobiasBora replied:
fixed in https://github.com/NixOS/nixpkgs/pull/324552

K900 commented on pkgs/kde/gear/kdenlive/default.nix:
```diff
@@ -31,7 +34,12 @@ mkKdeDerivation {
-  ];
+  ] ++ lib.optional enableGlaxnimate (
+    substituteAll {
+      src = ./dependency-paths-glaxnimate.patch;
```
Because this patch is separate (again, why?)

tobiasBora replied:
fixed in https://github.com/NixOS/nixpkgs/pull/324552

K900 commented on pkgs/kde/gear/kdenlive/default.nix:
```diff
@@ -43,7 +51,7 @@ mkKdeDerivation {
     shared-mime-info
     libv4l
-  ];
+  ] ++ lib.optional enableGlaxnimate glaxnimate;
```
And this is still optional (again, why?)

tobiasBora replied:
sorry, my mistake, fixed in https://github.com/NixOS/nixpkgs/pull/324552

tobiasBora commented:
> Respectfully, what the hell? How did this get merged in the current state?

Sorry, I did made a mistake in the PR by making glaxnimate only optional for one Qt version, and given that I addressed your last comment I thought my changes would certainly be good enough, but I was certainly a bit too much in a hurry.

And I did not explicitly asked to merge, I just asked what would happen now since originally turion contacted me to just rebase it and I did it ^^' But I'm wondering if it would make sense to create a bot that, when requested, can automatically merge a PR after X days and warn other PR that a PR will be merged soon if it conflicts. This way, it allows a bit of time for the other maintainers to catchup on the changes, without having this annoying "let's wait the answer of XXX", that finally never arrives, and stuck the PR forever until a merging conflict is solved…

Also, even if reaching the perfect code is definitely great, I think we need to find a balance between perfect code and reasonable time lost on improving the code, and compare it with the added value of this cleaning: this can be very time and energy (I mean personal energy but actually also in CPU power ^^) consuming, as any refactoring needs full recompilations, tests, not only from the contributor, but also from all reviewers, and each reviewing round increases the changes for the PR to be abandoned since any tiny change can turn into a massive lost of time. 

> The implementation isn't consistent across Qt versions.

You mean because `glaxnimate` is optional in one qt version and not the other? That's my mistake, I thought I fixed it in the two and cause I was a bit busy I overlooked it.

> none of the maintainers signed off on this

What do you mean? @turion is a maintainer of kdenlive, and except from you, no one got active since @turion contacted me.

> and now we'll have to do a follow-up PR because I don't want to have this in tree in this state.

Done in https://github.com/NixOS/nixpkgs/pull/324552

turion commented:
I agree with all of the above.

@K900 Apologies that the state of the PR didn't meet your expectations. It is over a year old, has been rebased, approved, commented on, people dropped by asking for changes and then were unresponsive for a long time. I found it frustrating that while tobiasBora has put effort in this, been patient, and created value for the community, there was always something that held the PR back. So I decided to go forward in an imperfect state. Yes, I could have waited for your extensive and detailed feedback that certainly improves the code quality. But I've also seen enough PRs that slowly wither because there is review after review burning out the contributor. I perceived losing tobiasBora here, and I didn't want to. 

> none of the maintainers signed off on this

I'm the sole "maintainer" of the old kdenlive (I put quotes here because I'm not proficient enough in Qt to do any nontrivial maintenance like tobiasBora has done). I have no idea how to identify the correct maintainer of the plasma 6 version, of which I was completely unaware until recently. It seems that a group of people is now maintaining all of the kdePackages at the same time? Should I have pinged all of them?

K900 commented:
Generally, the KDE team maintains all KDE 6 packages as a baseline. At least the committers on the team are pinged automatically by CODEOWNERS.
